### PR TITLE
Add container image build to repo

### PR DESCRIPTION
**Pull request by lvlie**
This PR adds a Dockerfile and example docker-compose.yaml to allow for easier administration. 

In order to enable the build in your repo, you should:

* create a PAT for your account (User > Settings > Developer settings > Personal Access Tokens) with the read and write package rights
* put the token into a secret variable for this repo (Settings > secrets > New secret). Give it the name "PACKAGE_TOKEN" as this is configured in the github action
* Change the visibility of the created "package" (User > Your Profile > Packages > pycryptobot/pycryptobot > Package settings > Change visibility

It is a bit of work, but it makes usability a lot easier :) 

I also fixed all markdown-lint violations in the readme file.

You can check the actual container that was built out of this: ghcr.io/lvlie/pycryptobot/pycryptobot:latest

And I think the "action" log should also be visible as well: https://github.com/lvlie/pycryptobot/runs/2307818414?check_suite_focus=true

### CODE REVIEW

**Review by whittlem — APPROVED** [review]

**Comment by whittlem** [comment]
Thank you for your contribution.

**Comment by whittlem** [comment]
I'm not using the Enterprise version of Github. I don't have a "Personal Access Tokens" option in the Settings. Is there an equivalent version for the public version? Would not adding the "Personal Access Token" into an environment variable pose a security risk?

**Comment by lvlie** [comment]
The PAT option is under developer settings in the settings page and is not related to any paid version afaik, I'm not a paying user as well :-). 
The secret should only be available to you as owner of the repo, and only has access to your packages (limit the rights the PAT has). Also, in github actions the value is not visible in the current workflow (you can check the output in my fork) and would always be redacted even if outputted. 
That does not mean there are no ways around this and you should take care not to merge github actions content that do you did not review. 

Btw, i see the badge is not working at the moment, that's because I changed the workflow name. Should be an easy fix

File: .github/workflows/container.yml [hunk 1]
```diff
@@ -1,10 +1,11 @@
-name: Docker
+name: Container
```
Comment (lvlie) [inline .github/workflows/container.yml:1]: This made the badge not work...

File: README.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
+[![Docker](https://github.com/whittlem/pycryptobot/actions/workflows/docker-publish.yml/badge.svg)](https://github.com/whittlem/pycryptobot/actions/workflows/docker-publish.yml)
```
Comment (lvlie) [inline README.md:1]: If you change "docker-publish" to "container" (twice) the badge will work again

**Review by whittlem — COMMENTED** [review]
I've done the first two points as per your instructions...

* create a PAT for your account (User > Settings > Developer settings > Personal Access Tokens) with the read and write package rights

* put the token into a secret variable for this repo (Settings > secrets > New secret). Give it the name "PACKAGE_TOKEN" as this is configured in the github action

* Change the visibility of the created "package" (User > Your Profile > Packages > pycryptobot/pycryptobot > Package settings > Change visibility

I'm not sure what to do on the last point. I found the "Packages" section but it says I should  "Choose an ecosystem". Should a package have been created or do I still need to do something here?

**Comment by lvlie** [comment]
Ah yeah, the github action should be run first (which requires the PAT to be configured through 1 and 2), before you can change the visibility (point 3) indeed. It should be created for any push you do to the main branch or you should be able to re-run the last one (that failed because PACKAGE_TOKEN was not set) here somewhere: https://github.com/whittlem/pycryptobot/actions/runs/740528011

**Comment by whittlem** [comment]
When I run the "Push" under "Actions" I see this...

Run echo "" | docker login ghcr.io -u whittlem --password-stdin
  echo "" | docker login ghcr.io -u whittlem --password-stdin
  shell: /usr/bin/bash -e {0}
  env:
    IMAGE_NAME: pycryptobot
Error: Cannot perform an interactive login from a non TTY device
Error: Process completed with exit code 1.

**Comment by whittlem** [comment]
It got to the 4th step... failed on "Log into registry"

**Comment by whittlem** [comment]
I see the Docker image is building with Python38. It should be Python39. Can you update that?

**Comment by lvlie** [comment]
> It got to the 4th step... failed on "Log into registry"

It should be something like:
image

Are you sure the PACKAGE_TOKEN secret var as needed [here](https://github.com/whittlem/pycryptobot/blob/main/.github/workflows/container.yml#L35) is set correctly?

**Comment by whittlem** [comment]
I clicked "Re-run jobs" and it went past "Log into registry" now.

It seems to complete but there is an error on the "Push image" step.

Run IMAGE_ID=ghcr.io/whittlem/pycryptobot/$IMAGE_NAME
IMAGE_ID=ghcr.io/whittlem/pycryptobot/pycryptobot
VERSION=latest
The push refers to repository [ghcr.io/whittlem/pycryptobot/pycryptobot]
6f8609d24ef6: Preparing
2fa061ffe147: Preparing
bbb8bb5a0fd5: Preparing
8838b2c54cd7: Preparing
41963ce1cb78: Preparing
denied: failed_precondition: Improved container support has not been enabled for 'whittlem'. Learn more: https://docs.github.com/packages/getting-started-with-github-container-registry/enabling-improved-container-support
Error: Process completed with exit code 1.

Any ideas?

**Comment by lvlie** [comment]
> I see the Docker image is building with Python38. It should be Python39. Can you update that?

Yeah I'm basing this of the free version of RHEL for containers, which does not yet have 3.9 (RHEL is enterprise-y, which should mean it is one of the more secure options). It seems to work correctly for 3.8 as well and that should be supported for a long time as well. 

Options are:

* wait for when Red Hat releases 3.9
* move to another container supplying python already (options enough, like the upstream python images on docker hub; the only reason I did not use that is because afaik, docker hub now wants you to login if you dont want to hit pull quotas and that would require another secret)

**Comment by lvlie** [comment]
> I clicked "Re-run jobs" and it went past "Log into registry" now.
> 
> It seems to complete but there is an error on the "Push image" step.
> 
> **
> 
> Any ideas?

Ah, right! That is something I enabled some time ago. Follow the steps in the link and it should work :)

**Comment by whittlem** [comment]
Ok, thanks. I'll try it.

**Comment by whittlem** [comment]
So that worked. I've not done this before with Github.

Does that mean people can install and run the app something like?

docker run --rm pycrptobot

**Comment by whittlem** [comment]
I see you have updated the README.md which answered my question.

    % docker pull ghcr.io/whittlem/pycryptobot:latest
    % docker pull ghcr.io/whittlem/pycryptobot/pycryptobot:latest
    
That is pretty amazing. I think people will love this. Does it run this workflow every time I push an update?

And you sure everything works fine with Python 3.8.x? I thought for Binance it needed Python 3.9.x.

**Comment by lvlie** [comment]
Correct! It will create a new container image for every push :) 

Because the image will only contain this repo (and no config), users will need to mount in their config.json with the -v flag (`-v ./config.json:/app/config.json` if their config.json is in the current working dir) and since pycryptobot.py is the `entrypoint` they can pass parameters to pycryptobot on the end of the command line. If you want to troubleshoot, you can also override the entrypoint with bash (like `docker run -it --rm --entrypoint bash ghcr.io/whittlem/pycryptobot/pycryptobot:latest`) and go from there. If they also mount in the config.json they can edit it there as well. 

Another cool example is in the docker-compose.yaml: you can start all your bots with multiple container specifications. The only limitations being:
* all containers should have a different name 
* all containers should mount a different config.json
* you should stay within the limits on the api :D

3.8 seems to work fine for binance for me. Not seeing any errors at least :)

**Comment by whittlem** [comment]
This is a great addition. Thanks very much.

**Comment by lvlie** [comment]
Btw, while we're doing docker 101 here, you can of course run the container with `docker-compose up -d` or `docker run -d ...` in order to spin it up as a daemon. You can use `docker exec -it <containername> bash` to get into the container and troubleshoot, read logs, etc. 

Also, in the github action workflow, you can also first schedule all kinds of tests on your code and make the container build dependent on that. Havent worked on github actions much (or write a lot of python for that matter), so wont be able to help much there. 

Anyway, good luck :)

**Comment by whittlem** [comment]
That's great, thanks.

**Pull request merged**